query-parser/src/parse.rs

```diff
@@ -70,6 +69,130 @@ macro_rules! matches_plain_symbol {
     }
 }
 
+trait FromValue<T> {
+    fn from_value(edn::Value) -> Option<T>;
+}
+
+impl FromValue<PatternNonValuePlace> for PatternNonValuePlace {
+    fn from_value(v: edn::Value) -> Option<PatternNonValuePlace> {
+        match v {
+            edn::Value::Integer(x) => if x >= 0 {
+                Some(PatternNonValuePlace::Entid(x as u64))
+            } else {
+                None
+            },
+            edn::Value::PlainSymbol(x) => if x.0.as_str() == "_" {
+                Some(PatternNonValuePlace::Placeholder)
+            } else if x.is_var_symbol() {
+                Some(PatternNonValuePlace::Variable(Variable(x)))
+            } else {
+                None
+            },
+            edn::Value::NamespacedKeyword(x) =>
+                Some(PatternNonValuePlace::Ident(x)),
+            _ => None,
+        }
+    }
+}
+
+impl FromValue<PatternValuePlace> for PatternValuePlace {
```

Comment:
So, the `serde` way of doing these conversions is fancy: I think they generate serializing/deserializing code statically, strong typing these things.  Given how frequent `edn::Value` <-> domain object is going to be, maybe which should learn their secrets and make this first class?